src/rust/lib/fuzzly/src/metric.rs

```diff
@@ -0,0 +1,139 @@
+// =============
+
+/// Provides functions for measure pattern match score for specific text from various aspect.
+///
+/// The pattern match is represented as a path in `SubsequenceGraph` (see it's docs for details).
+/// Its score is counted as a sum of measures "how good is the vertex/edge" for each vertex and
+/// edge on the path.
+pub trait Metric {
+    /// How good is vertex on the path on the Subsequence Graph.
+    fn measure_vertex(&self, vertex:subsequence_graph::Vertex, text:&str, pattern:&str) -> f32;
+
+    /// How good is the edge on the path on the Subsequence Graph.
+    fn measure_edge(&self, edge:subsequence_graph::Edge, text:&str, pattern:&str) -> f32;
+
+    /// Return a new metric being a sum of this and `rhs`.
+    fn sum<Rhs:Metric>(self, rhs:Rhs) -> Sum<Self,Rhs> where Self:Sized { Sum(self, rhs) }
+}
+
+
+
+// ==========================
+// === The Default Metric ===
+// ==========================
+
+/// The default metric, recommended by this library
+pub fn default() -> impl Metric {
+    SubsequentLettersBonus::default().sum(CaseMatchBonus::default())
+}
+
+
+
+// =======================
+// === Implementations ===
+// =======================
+
+// === Sum ===
+
+/// The structure representing the sum of two metrics
+#[derive(Copy,Clone,Debug,Default)]
+pub struct Sum<Metrics1,Metrics2>(Metrics1,Metrics2);
+
+impl<M1:Metric, M2:Metric> Metric for Sum<M1,M2> {
+    fn measure_vertex(&self, vertex:subsequence_graph::Vertex, text:&str, pattern:&str) -> f32 {
+        let Self(left,right) = self;
+        let left             = left.measure_vertex(vertex,text,pattern);
+        let right            = right.measure_vertex(vertex,text,pattern);
+        left + right
+    }
+
+    fn measure_edge(&self, edge:subsequence_graph::Edge, text:&str, pattern:&str) -> f32 {
+        let Self(left,right) = self;
+        let left             = left.measure_edge(edge,text,pattern);
+        let right            = right.measure_edge(edge,text,pattern);
+        left + right
+    }
+}
+
+
+// === SubsequentLettersBonus ===
+
+/// A metric which measure how far are matched letters from each other and how far is first matched
+/// char from text beginning and last character from text ending.
+#[derive(Copy,Clone,Debug)]
+pub struct SubsequentLettersBonus {
+    /// The base weight of this metric.
+    pub base_weight:f32,
+    /// How important is the distance of first matched char from the text beginning.
+    pub beginning_weight:f32,
+    /// How important is the distance of last matched char from the text ending.
+    pub ending_weight:f32,
+}
+
+impl Default for SubsequentLettersBonus {
+    fn default() -> Self {
+        SubsequentLettersBonus {
+            base_weight      : 1.0,
+            beginning_weight : 0.5,
+            ending_weight    : 0.01,
+        }
+    }
+}
+
+impl Metric for SubsequentLettersBonus {
+    fn measure_vertex(&self, vertex:subsequence_graph::Vertex, text: &str, _pattern: &str) -> f32 {
```

Comment:
why the pattern is not used here and below?

Reply:
Because this bonus rely only on the letters position, not what the characters actually are.